_posts/2021-09-20-where-the-store-is-stored.md

```diff
@@ -0,0 +1,274 @@
+drilling](https://sebhastian.com/react-prop-drilling/). To overcome this, we
+can add a `Context` to make our state and its API available for all child
+components via a simple `useContext` - even if [it's not made for
+that](https://twitter.com/sebmarkbage/status/1219836431972978689). The idea is
+the same: the store is stored inside React "runtime" via `useState`.
+
+## Programatic store
+
+And finally, the good old way: dependency injection by hand. Let's say we're in
+a full custom app (no React and stuff) and we don't want to use modules. We can
+then craft a store instance when the app starts and pass it along everywhere via
+function parameters.
+
+Let's write a store factory very similar to the ES6 module option.
+
+```js
+// store.js
+function buildStore() {
+  const _cars = [];
+
+  return {
+    add: (car) => {
+      _cars.push(car);
+    },
+
+    list: () => {
+      return _cars;
+    },
+  }
+}
+```
+
+Now that our store's factory is ready, we just have to craft a store instance
+sowhere in app's entrypoint and pass it along everywhere.
+
+```js
+// index.js
+
+function initApp() {
+  const store = buildStore();
+
+
+  startApp(store);
+}
+
+initApp();
+```
+
+```js
+function startApp(store) {
+  store.add('new car');
+
+  document.body.innerText = store.list().join(', ');
+
+  // ...
+}
+```
+
+Of course, any function that needs to access the store will have to receive it
+via its parameters and we're back with our props drilling issue.
+
+# Pros and Cons of each store location 
+
+Ok so three potential store locations. You can see me coming with the list of
+pros and cons for each of them. 
+
+Let's start with modules. Pros: it's raw ES6 (`import` / `export`) so quite simple
+to grasp. It's also very flexible. We can do more or less whathever we what in a
+module and every part of our codebase can use it.
+
+But there is a big constraint. Not all codebases always run in the same runtime.
+Actually a lot of single page applications run in both browser and NodeJS
+runtimes, where the contexts are very different. In the browser, we can make the
+assumption that there is only one user at a given time using the runtime. So our
+module state will be "our user state". On NodeJS, we will probably use the code
+to render the page server-side for multiple http requests. In this situation,
+our module state will be shared between mutliple. Boom! Data leak security
+issue.
+
+For librairy or framework locations likes our React `useState`, the big
+advantage is probably the strong integration with the corresponding library.
+Even bigger store solutions like Redux and Cie come with their binding like
+`react-redux` that does the job for us. No need to handle the Context / Provider
+stuff I mentioned. No issue with server-side rendering: the store being "in
+React runtime" it will be scoped to the `ReactDOM.render` or
+`ReactDOMServer.renderToString` (which will be called once for every http
+request).
```

Comment:
```suggestion
For library or framework locations such as our React `useState`, the big
advantage is probably the strong integration with the corresponding library.
Even bigger store solutions like Redux come with their binding like
`react-redux` that does the job for us. No need to handle the Context / Provider
stuff I mentioned. No issue with server-side rendering: the store being "in
React runtime" it will be scoped to the `ReactDOM.render` or
`ReactDOMServer.renderToString` (which will be called once for every http
request).
```